src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapPartialResultsProcessor.cs

```diff
@@ -335,29 +340,52 @@ private static void AddResult(SearchResponse partialResults, SearchResponse newR
 
     internal sealed class PartialResultsRetriever
     {
+        private static readonly TimeSpan s_maxThreadIdleTime = TimeSpan.FromSeconds(30);
```

Comment:
The idea is to start the thread when necessary, by calling `StartIfNotRunning`. Then keep it alive for a while (the value of 30 seconds is totally arbitrary), and let the worker thread finish gracefully if there's nothing more to do.

It was [suggested](https://github.com/dotnet/runtime/issues/75614#issuecomment-1562175421) that compatibility with the current stable API is important, so that I decided not to rewrite the whole internal logic. [A fully `Task` based async `LdapConnection`](https://github.com/dotnet/runtime/issues/75614#issuecomment-1648628200) requires quite significant changes to the external API.

Reply:
> Then keep it alive for a while, and let the worker thread finish gracefully if there's nothing more to do.

How this make sure that there is nothing more to do? For me it looks there is no guarantee that all `AsyncResults` in the `_resultList` could be retrieved/processed within these 30 seconds.

Reply:
My understanding of the async processing could be wrong. This is how I understood the process:

All `AsyncResult` instances seem to be created by the  ".NET LDAP Results Retriever" thread in sequential order of the `_resultList`. The thread's loop is triggered by the `ManualResetEvent _workThreadWaitHandle`. The calls to `_workThreadWaitHandle.Set()`, `_workThreadWaitHandle.Reset()`, `_workThreadWaitHandle.WaitOne(s_maxThreadIdleTime)`, and  `_workThreadWaitHandle.WaitOne(0)` are synchronized with a `lock` on the same mutex. The main method that actually works on the result list,`RetrievingSearchResults()`, should only return once the current `_resultList` is processed, so there shouldn't be any outstanding requests/results when the `ThreadRoutine()` method is exited.

Reply:
> All AsyncResult instances seem to be created by the ".NET LDAP Results Retriever" thread in sequential order of the _resultList.

It looks to me LdapConnection.BeginSendRequest(...) creating those and adding to them to the `_resultList` via  s_partialResultsProcessor.Add(...) https://github.com/dotnet/runtime/blob/d7ae8c61f0937df4dcffe88ba96eae65fa523a78/src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs#L373-L374

which will not set the `_workThreadWaitHandle` in case there is a work to do.

All access to `_resultList` is locked yes, but I believe those locks will not guarantee that all methods run in specific order, I mean it will not stop s_partialResultsProcessor.Add(...) operation could run 2+ times before background thread start processing with `RetrievingSearchResults()` even though it could happen very rarely.

> The calls to `_workThreadWaitHandle.Set()`, `_workThreadWaitHandle.Reset()`, `_workThreadWaitHandle.WaitOne(s_maxThreadIdleTime)`, and  `_workThreadWaitHandle.WaitOne(0)` are synchronized with a `lock` on the same mutex. The main method that actually works on the result list,`RetrievingSearchResults()`, should only return once the current `_resultList` is processed, so there shouldn't be any outstanding requests/results when the `ThreadRoutine()` method is exited.

Sorry I don't follow (see) this, but I am not expert in threading area and do not know this legacy code, I'll see what experts say